telemetry/Tracks/class-tracks-client.php

```diff
@@ -0,0 +1,187 @@
+	 * @return bool|WP_Error True if the event was enqueued for recording.
+	 *                       False if the event is not recordable.
+	 *                       WP_Error if the event is generating an error.
+	 */
+	public function record_event_asynchronously( Tracks_Event $event ) {
+		$is_event_recordable = $event->is_recordable();
+
+		if ( true !== $is_event_recordable ) {
+			return $is_event_recordable;
+		}
+
+		static::instance()->events[] = $event;
+
+		return true;
+	}
+
+	/**
+	 * Records an event synchronously (using a GET request).
+	 *
+	 * @param Tracks_Event $event The event to record.
+	 * @return bool|WP_Error True if recording the event succeeded.
+	 *                       False if the event is not recordable.
+	 *                       WP_Error if any error occurred.
+	 */
+	public function record_event_synchronously( Tracks_Event $event ) {
+		$is_event_recordable = $event->is_recordable();
+
+		if ( true !== $is_event_recordable ) {
+			return $is_event_recordable;
+		}
+
+		$pixel_url = static::instance()->generate_pixel_url( $event );
+
+		if ( null === $pixel_url ) {
+			return new WP_Error(
+				'invalid_pixel',
+				'cannot generate tracks pixel for given input',
+				array( 'status' => 400 )
+			);
+		}
+
+		// Add the Request Timestamp and URL terminator just before the HTTP
+		// request.
+		$pixel_url .= '&_rt=' . static::build_timestamp() . '&_=_';
+
+		$request = wp_safe_remote_get(
+			$pixel_url,
+			array(
+				'blocking'    => false,
+				'redirection' => 2,
+				'httpversion' => '1.1',
+				'timeout'     => 1,
+			)
+		);
+
+		if ( is_wp_error( $request ) ) {
+			return $request;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Outputs a Tracks pixel for every registered event.
+	 */
+	public function render_tracking_pixels(): void {
+		foreach ( $this->events as $event ) {
+			$pixel_url = $this->generate_pixel_url( $event );
+
+			if ( null === $pixel_url ) {
+				continue;
+			}
+
+			echo '<img style="position: fixed;" src="', esc_url( $pixel_url ), '" />';
+		}
+
+		$this->events = array();
+	}
+
+	/**
+	 * Generates a Tracks pixel URL that will record an event when accessed.
+	 *
+	 * @param Tracks_Event $event The event for which to generate the pixel URL.
+	 * @return ?string The pixel URL or null if an error occurred.
+	 */
+	protected function generate_pixel_url( Tracks_Event $event ): ?string {
+		$event_data = $event->get_data();
+
+		if ( is_wp_error( $event_data ) ) {
+			return null;
+		}
+
+		$args = get_object_vars( $event_data );
+
+		// Request Timestamp and URL Terminator must be added just before the
+		// HTTP request or not at all.
+		unset( $args['_rt'], $args['_'] );
+
+		return esc_url_raw(
+			static::PIXEL_BASE_URL . '?' . http_build_query( $args )
+		);
+	}
+
+	/**
+	 * Records any remaining events synchronously.
+	 */
+	public function record_remaining_events(): void {
+		foreach ( $this->events as $event ) {
+			static::instance()->record_event_synchronously( $event );
+		}
```

Comment:
The REST API supports bulk events which would be better since we could just make one remote request instead of multiples: https://github.com/Automattic/vip-cli/blob/2a8c90afa9ed2e5d784e820c5da20183094b6711/src/lib/analytics/clients/tracks.ts#L84-L105

Reply:
Oh good idea! I'll look into that 👍

Reply:
Changed in 7ae681a528aaf3678c307e68e0c56c7f134b2748